shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,177 @@
+    }
+  }
+
+  late final bool _isAvailable;
+
+  final Updater _updater;
+
+  final IsolateRun _run;
+
+  @override
+  bool get isAvailable => _isAvailable;
+
+  @override
+  Future<Patch?> readCurrentPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.currentPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<Patch?> readNextPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.nextPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<UpdateStatus> checkForUpdate() async {
+    if (!_isAvailable) return UpdateStatus.unavailable;
+
+    final isUpdateAvailable = await _run(_updater.checkForUpdate);
+    if (isUpdateAvailable) return UpdateStatus.outdated;
+
+    final (current, next) = await (readCurrentPatch(), readNextPatch()).wait;
+    return next != null && current?.number != next.number
+        ? UpdateStatus.restartRequired
+        : UpdateStatus.upToDate;
+  }
+
+  @override
+  Future<void> update() async {
+    if (!_isAvailable) return;
+
+    Pointer<UpdateResult> result = nullptr;
+
+    try {
+      result = await _run(_updater.update);
+    } catch (_) {
+      return _legacyFallback();
+    }
+
+    try {
+      if (result == nullptr) {
+        const reason = UpdateFailureReason.unknown;
+        final message = reason.toFailureMessage();
+        throw UpdateException(reason: reason, message: message);
+      }
+
+      final status = result.ref.status;
+
+      if (status == SHOREBIRD_UPDATE_INSTALLED) return;
+
+      final reason = status.toFailureReason();
+      final details = result.ref.message != nullptr
+          ? result.ref.message.cast<Utf8>().toDartString()
+          : 'unknown';
+      final message = reason.toFailureMessage(details);
+      throw UpdateException(message: message, reason: reason);
+    } finally {
+      _updater.freeUpdateResult(result);
+    }
+  }
+
+  // Fallback to downloadUpdate if update is not available.
+  Future<void> _legacyFallback() async {
+    await _run(_updater.downloadUpdate);
+    final (current, next) = await (readCurrentPatch(), readNextPatch()).wait;
+    final status = next != null && current?.number != next.number
+        ? UpdateStatus.restartRequired
+        : UpdateStatus.upToDate;
+    if (status == UpdateStatus.restartRequired) return;
+    throw const UpdateException(
+      message: '''
+Downloading update failed but reason is unknown due to legacy updater.
+Please upgrade the Shorebird Engine for improved error messages.''',
```

Comment:
```suggestion
Please upgrade to Flutter 3.24.4 or later for improved error messages.''',
```

Reply:
I think being even more specific here will be useful once we know which engine rev will contain this. As it stands, if someone has already cut a release with 3.24.4, they won't be able to use this functionality even if they patch the release with the updated `package:shorebird_code_push` api.

Reply:
Sure happy to adjust this once we have the new engine revision